third_party/txt/src/txt/paragraph_txt.cc

```diff
@@ -354,11 +355,40 @@ bool ParagraphTxt::ComputeLineBreaks() {
 
     size_t breaks_count = breaker_.computeBreaks();
     const int* breaks = breaker_.getBreaks();
-    for (size_t i = 0; i < breaks_count; ++i) {
-      size_t break_start = (i > 0) ? breaks[i - 1] : 0;
-      size_t line_start = break_start + block_start;
-      size_t line_end = breaks[i] + block_start;
-      bool hard_break = i == breaks_count - 1;
+    for (size_t i = 0; i < breaks_count && !early_ellipsized; ++i) {
+      size_t break_start;
+      size_t line_start;
+      size_t line_end;
+      bool hard_break;
+      double line_width = breaker_.getWidths()[i];
+      // We only want to early ellipsized if user choose ellipsizing by character.
+      // Other type of ellipsizing should still parse through the paragraph.
+      if (paragraph_style_.ellipsized_by_character &&
+          i < breaks_count - 1 &&
+          (paragraph_style_.unlimited_lines() || line_ranges_.size() == paragraph_style_.max_lines - 1)) {
+        break_start = (i > 0) ? breaks[i - 1] : 0;
+        line_start = break_start + block_start;
+        // The width of current line is shorter than constraint. If we want to pack
+        // in as many characters as we can, we have to borrow more characters from
+        // next line.
+        //
+        // i < breaks_count - 1 implies there is at least one more line after, and
+        // it is guaranteed the width of current line + next line will exceed the
+        // width constraint; otherwise, they will not break into different lines
+        // in the first place.
+        line_end = breaks[i + 1] + block_start;
+        hard_break = true;
```

Comment:
I think this should be:
`hard_break = (i + 1) == (breaks_count - 1)`

hard_break should only be true if the merged line ends in a newline character

Reply:
I think this won't matter if we decide to not keep two implementations. I set it to true this line will be the last line after compute line break, and I am not sure if there is anything that depends on the last line must end with hard break. We also do similar thing that insert a hard break at the end of paragraph.

Reply:
The code that will be affected is the calculation of `line_end_including_newline`.  `line_end_including_newline` should only differ from `line_end` if the line contains a newline character.

Reply:
It is also used to determine if it need to justify the line, but since this is the last line, justify will not apply anyway. It really doesn't matter this is true or false at this moment, but given my above justification, do you still think this should be false?

Reply:
Yes - I think `hard_break` should be false so that `line_end_including_newline` will be accurate